base/math.jl

```diff
@@ -547,8 +547,8 @@ julia> hypot(3, 4im)
 ```
 """
 hypot(x::Number, y::Number) = hypot(promote(x, y)...)
-hypot(x::Complex, y::Complex) = hypot(promote(abs(x),abs(y))...)
-hypot(x::Integer, y::Integer) = hypot(promote(float(x), float(y))...)
+hypot(x::Complex, y::Complex) = hypot(abs(x), abs(y))
+hypot(x::Integer, y::Integer) = hypot(float(x), float(y))
```

Comment:
I think we should actually have  `hypot(x::T, y::T) where {T<:Number} = hypot(float(x), float(y)`   and removing the `Integer` method case.

Reply:
Won't that break bigfloats et. al?

Reply:
Right...

Reply:
No it won't actually. `hypot` has specialized methods in MPFR.

Reply:
To me this still feels fishy as it means hypot will convert custom number types into floats if they don't specifically define a `hypot` function.

Reply:
We already define `mathfunction(x::Real) = exp(float(x))` for numerous elementary math functions.

Reply:
In any case we should probably change it from `T<:Number` to `T<:Real` .